File: javascript/smartlist/smartlist-api-mongodb-project/src/modules/report/report.assessment.controller.ts
import {
  Controller,
  Get,
  Param,
  HttpCode,
  ParseIntPipe,
  Query,
} from '@nestjs/common';
import { ApiOperation, ApiResponse } from '@nestjs/swagger';
import { ReportAssessmentService } from './report.assessment.service';
import { AssessmentBlockCandidatePerformanceRespDto } from './dto/response/assessment-block-candidate-performance.resp.dto';
import { AssessmentCandidatePerformanceRespDto } from './dto/response/assessment-candidate-performance.resp.dto';
import { AssessmentStatusSummaryRespDto } from './dto/response/assessment-status-summary.resp.dto';
import { AssessmentBlockAverageScoreRespDto } from './dto/response/assessment-block-average-score.resp.dto';
import { ReportResponseCodes } from './report.response.codes';
import { RateSummaryRespDto } from './dto/response/rate-summary.resp.dto';
import { HighLevelSummaryRespDto } from './dto/response/highlevel-summary.resp.dto';
import { TimeSeriesSummaryRespDto } from './dto/response/time-series-summary.resp.dto';
import { Role } from 'src/enums/role';
import { Roles } from 'src/guards/roles.decorator';
import { ReportDocumentRepository } from './report.document.repository';

@Controller('report/')
@Roles(Role.SUPERADMIN, Role.ADMIN, Role.MEMBER)
export class ReportAssessmentController {
  constructor(
    private readonly reportAssessmentService: ReportAssessmentService,

    private reportDocumentRepository: ReportDocumentRepository,
  ) { }


  @Get('test/organization/:organizationId/assessment/:assessmentId')
  async test(
    @Param('organizationId') organizationId: string,
    @Param('assessmentId') assessmentId: string,
    @Query('from') from: Date,
    @Query('to') to: Date,
  ): Promise<number> {
    return await this.reportDocumentRepository.getCountCompleted(
      organizationId,
      assessmentId,
      from,
      to,
    );
  }

  @ApiOperation({
    summary:
      'Get high level summary by organization, assessment and date range',
  })
  @ApiResponse({
    ...ReportResponseCodes.SUCCESS,
    type: [HighLevelSummaryRespDto],
  })
  @ApiResponse(ReportResponseCodes.BAD_REQUEST)
  @ApiResponse(ReportResponseCodes.INVALID_ASSESSMENT_ID)
  @ApiResponse(ReportResponseCodes.INVALID_DATE)
  @HttpCode(200)
  @Get('assessment/:assessmentId/highlevel/summary')
  async getAssessmentHighLevelSummary(
    @Param('assessmentId') assessmentId: string,
    @Query('from') from: Date,
    @Query('to') to: Date,
  ): Promise<HighLevelSummaryRespDto> {
    return await this.reportAssessmentService.getHighLevelSummary(
      assessmentId,
      from,
      to,
    );
  }

  @ApiOperation({
    summary: 'Get rate summary for organization and assessment by date range',
  })
  @ApiResponse({
    ...ReportResponseCodes.SUCCESS,
    type: [RateSummaryRespDto],
  })
  @ApiResponse(ReportResponseCodes.BAD_REQUEST)
  @ApiResponse(ReportResponseCodes.INVALID_ASSESSMENT_ID)
  @ApiResponse(ReportResponseCodes.INVALID_DATE)
  @HttpCode(200)
  @Get('assessment/:assessmentId/rate/summary')
  async getAssessmentRateSummary(
    @Param('assessmentId') assessmentId: string,
    @Query('from') from: Date,
    @Query('to') to: Date,
  ): Promise<RateSummaryRespDto> {
    return await this.reportAssessmentService.getRateSummary(
      assessmentId,
      from,
      to,
    );
  }

  @ApiOperation({ summary: 'Get assessment status summary count' })
  @ApiResponse({
    ...ReportResponseCodes.SUCCESS,
    type: [AssessmentStatusSummaryRespDto],
  })
  @ApiResponse(ReportResponseCodes.BAD_REQUEST)
  @ApiResponse(ReportResponseCodes.INVALID_ASSESSMENT_ID)
  @ApiResponse(ReportResponseCodes.INVALID_DATE)
  @HttpCode(200)
  @Get('assessment/:assessmentId/status/summary')
  async getAssessmentStatusSummary(
    @Param('assessmentId') assessmentId: string,
    @Query('from') from: Date,
    @Query('to') to: Date,
  ): Promise<AssessmentStatusSummaryRespDto> {
    return await this.reportAssessmentService.getAssessmentStatusSummary(
      assessmentId,
      from,
      to,
    );
  }

  @ApiOperation({
    summary: 'Get assessment block average score for assessment',
  })
  @ApiResponse({
    ...ReportResponseCodes.SUCCESS,
    type: [AssessmentBlockAverageScoreRespDto],
  })
  @ApiResponse(ReportResponseCodes.BAD_REQUEST)
  @ApiResponse(ReportResponseCodes.INVALID_ASSESSMENT_ID)
  @ApiResponse(ReportResponseCodes.INVALID_DATE)
  @HttpCode(200)
  @Get('assessment/:assessmentId/assessment-block/average-score')
  async getAssessmentBlockAverageScore(
    @Param('assessmentId') assessmentId: string,
    @Query('from') from: Date,
    @Query('to') to: Date,
  ): Promise<AssessmentBlockAverageScoreRespDto[]> {
    return await this.reportAssessmentService.getAssessmentBlockAverageScore(
      assessmentId,
      from,
      to,
    );
  }

  @ApiOperation({
    summary: 'Get assessment performance details for all candidates',
  })
  @ApiResponse({
    ...ReportResponseCodes.SUCCESS,
    type: [AssessmentCandidatePerformanceRespDto],
  })
  @ApiResponse(ReportResponseCodes.BAD_REQUEST)
  @ApiResponse(ReportResponseCodes.INVALID_ASSESSMENT_ID)
  @ApiResponse(ReportResponseCodes.INVALID_DATE)
  @HttpCode(200)
  @Get('assessment/:assessmentId/candidate/performance')
  async getAssessmentPerformanceDetailsForAllCandidates(
    @Param('assessmentId') assessmentId: string,
    @Query('from') from: Date,
    @Query('to') to: Date,
  ): Promise<AssessmentCandidatePerformanceRespDto[]> {
    return await this.reportAssessmentService.getAssessmentPerformanceForAllCandidates(
      assessmentId,
      from,
      to,
    );
  }

  @ApiOperation({
    summary: 'Get assessment block performance details for all candidates',
  })
  @ApiResponse({
    ...ReportResponseCodes.SUCCESS,
    type: [AssessmentBlockCandidatePerformanceRespDto],
  })
  @ApiResponse(ReportResponseCodes.BAD_REQUEST)
  @ApiResponse(ReportResponseCodes.INVALID_ASSESSMENT_BLOCK_ID)
  @ApiResponse(ReportResponseCodes.INVALID_DATE)
  @HttpCode(200)
  @Get('assessment-block/:assessmentBlockId/candidate/performance')
  async getAssessmentBlockPerformanceDetailsForAllCandidates(
    @Param('assessmentBlockId') assessmentBlockId: string,
    @Query('from') from: Date,
    @Query('to') to: Date,
  ): Promise<AssessmentBlockCandidatePerformanceRespDto[]> {
    return await this.reportAssessmentService.getAssessmentBlockPerformanceForAllCandidates(
      assessmentBlockId,
      from,
      to,
    );
  }

  @ApiOperation({
    summary: 'Time series data for assessment by date range',
  })
  @ApiResponse({
    ...ReportResponseCodes.SUCCESS,
    type: [TimeSeriesSummaryRespDto],
  })
  @ApiResponse(ReportResponseCodes.BAD_REQUEST)
  @ApiResponse(ReportResponseCodes.INVALID_ORGANIZATION_ID)
  @HttpCode(200)
  @Get('assessment/:assessmentId/timeseries/data')
  async getTimeSeriesSummaryForOrganization(
    @Param('assessmentId') assessmentId: string,
    @Query('from') from: Date,
    @Query('to') to: Date,
  ): Promise<TimeSeriesSummaryRespDto> {
    return await this.reportAssessmentService.getTimeSeriesSummaryForAssessment(
      assessmentId,
      from,
      to,
    );
  }
}

File: java/smartlist/src/main/java/com/smartlist/module/report/ReportAssessmentController.java
package com.smartlist.module.report;

import com.smartlist.module.report.dto.response.*;
import com.smartlist.utils.AppResponseCodes;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.media.ArraySchema;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.constraints.PastOrPresent;
import jakarta.validation.constraints.Positive;
import org.springframework.http.HttpStatus;
import org.springframework.security.access.annotation.Secured;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDate;
import java.util.List;

@RestController
@Tag(name = "Report-Assessment", description = "Report-Assessment API")
@RequestMapping(value ="/report/assessment/", produces = { "application/json" })
@ResponseStatus(HttpStatus.OK)
@Validated
@Secured({"SUPERADMIN", "ADMIN", "MEMBER"})
public class ReportAssessmentController {

    private final ReportAssessmentService reportAssessmentService;

    ReportAssessmentController(final ReportAssessmentService reportAssessmentService){
        this.reportAssessmentService = reportAssessmentService;
    }

    @Operation(summary = "Get high level summary by  assessment and date range")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", schema = @Schema(implementation = HighLevelSummaryRespDTO.class)) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = ReportResponseCodes.INVALID_DATE_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = ReportResponseCodes.INVALID_ASSESSMENT_ID_DESCRIPTION, content = @Content)
    })
    @GetMapping("{assessmentId}/highlevel/summary")
    public HighLevelSummaryRespDTO getAssessmentHighLevelSummary(
            @PathVariable("assessmentId") @Positive Integer assessmentId,
            @RequestParam("from") @PastOrPresent LocalDate from,
            @RequestParam("to") @PastOrPresent LocalDate to){
        return this.reportAssessmentService.getHighLevelSummary(assessmentId, from, to);
    }

    @Operation(summary = "Get rate summary for assessment by date range")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", schema = @Schema(implementation = RateSummaryRespDTO.class)) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = ReportResponseCodes.INVALID_DATE_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = ReportResponseCodes.INVALID_ASSESSMENT_ID_DESCRIPTION, content = @Content)
    })
    @GetMapping("{assessmentId}/rate/summary")
    public RateSummaryRespDTO getAssessmentRateSummary(
            @PathVariable("assessmentId") Integer assessmentId,
            @RequestParam("from") @PastOrPresent LocalDate from,
            @RequestParam("to") @PastOrPresent LocalDate to) {
        return  this.reportAssessmentService.getRateSummary(assessmentId, from, to);
    }

    @Operation(summary = "Get assessment status summary count")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", schema = @Schema(implementation = AssessmentStatusSummaryRespDTO.class)) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = ReportResponseCodes.INVALID_DATE_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = ReportResponseCodes.INVALID_ASSESSMENT_ID_DESCRIPTION, content = @Content)
    })
    @GetMapping("{assessmentId}/status/summary")
    public AssessmentStatusSummaryRespDTO getAssessmentStatusSummary(
            @PathVariable("assessmentId") Integer assessmentId,
            @RequestParam("from") @PastOrPresent LocalDate from,
            @RequestParam("to") @PastOrPresent LocalDate to) {
        return this.reportAssessmentService.getAssessmentStatusSummary(assessmentId, from, to);
    }

    @Operation(summary = "Get assessment block average score for assessment")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", array = @ArraySchema(schema = @Schema(implementation = AssessmentBlockAverageScoreRespDTO.class))) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = ReportResponseCodes.INVALID_DATE_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = ReportResponseCodes.INVALID_ASSESSMENT_ID_DESCRIPTION, content = @Content)
    })
    @GetMapping("{assessmentId}/assessment-block/average-score")
    public List<AssessmentBlockAverageScoreRespDTO> getAssessmentBlockAverageScore(
            @PathVariable("assessmentId") Integer assessmentId,
            @RequestParam("from") @PastOrPresent LocalDate from,
            @RequestParam("to") @PastOrPresent LocalDate to) {
        return  this.reportAssessmentService.getAssessmentBlockAverageScore(assessmentId, from, to);
    }

    @Operation(summary = "Get assessment performance details for all candidates")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", array = @ArraySchema(schema = @Schema(implementation = AssessmentCandidatePerformanceRespDTO.class))) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = ReportResponseCodes.INVALID_DATE_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = ReportResponseCodes.INVALID_ASSESSMENT_ID_DESCRIPTION, content = @Content)
    })
    @GetMapping("{assessmentId}/candidate/performance")
    public List<AssessmentCandidatePerformanceRespDTO> getAssessmentPerformanceDetailsForAllCandidates(
            @PathVariable("assessmentId") Integer assessmentId,
            @RequestParam("from") @PastOrPresent LocalDate from,
            @RequestParam("to") @PastOrPresent LocalDate to) {
        return this.reportAssessmentService.getAssessmentPerformanceForAllCandidates(assessmentId, from, to);
    }

    @Operation(summary = "Get assessment block performance details for all candidates")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", array = @ArraySchema(schema = @Schema(implementation = AssessmentBlockCandidatePerformanceRespDTO.class))) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = ReportResponseCodes.INVALID_DATE_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = ReportResponseCodes.INVALID_ASSESSMENT_BLOCK_ID_DESCRIPTION, content = @Content)
    })
    @GetMapping("assessment-block/{assessmentBlockId}/candidate/performance")
    public List<AssessmentBlockCandidatePerformanceRespDTO> getAssessmentBlockPerformanceDetailsForAllCandidates(
            @PathVariable("assessmentBlockId") Integer assessmentBlockId,
            @RequestParam("from") @PastOrPresent LocalDate from,
            @RequestParam("to") @PastOrPresent LocalDate to){
        return this.reportAssessmentService.getAssessmentBlockPerformanceForAllCandidates(assessmentBlockId, from, to);
    }

    @Operation(summary = "Time series data for assessment by date range")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", schema = @Schema(implementation = TimeSeriesSummaryRespDTO.class)) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = ReportResponseCodes.INVALID_ASSESSMENT_ID_DESCRIPTION, content = @Content)
    })
    @GetMapping("{assessmentId}/timeseries/data")
    public TimeSeriesSummaryRespDTO getTimeSeriesSummaryForOrganization(
            @PathVariable("assessmentId") Integer assessmentId,
            @RequestParam("from") @PastOrPresent LocalDate from,
            @RequestParam("to") @PastOrPresent LocalDate to) {
        return this.reportAssessmentService.getTimeSeriesSummaryForAssessment(assessmentId, from, to);
    }


}
